subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2031 @@
+			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+static struct bt_hfp_ag bt_hfp_ag_pool[CONFIG_BT_MAX_CONN];
+
+static struct bt_hfp_ag_cb *bt_ag;
+
+/* Sent but not acknowledged TX packets with a callback */
+static struct bt_ag_tx ag_tx[CONFIG_BT_HFP_AG_TX_BUF_COUNT * 2];
+static K_FIFO_DEFINE(ag_tx_free);
+static K_FIFO_DEFINE(ag_tx_notify);
+
+struct k_thread ag_thread;
+static K_KERNEL_STACK_MEMBER(ag_thread_stack, CONFIG_BT_HFP_AG_THREAD_STACK_SIZE);
+static k_tid_t ag_thread_id;
+
+static void bt_hfp_ag_set_state(struct bt_hfp_ag *ag, bt_hfp_state_t state)
+{
+	ag->state = state;
+
+	switch (state) {
+	case BT_HFP_DISCONNECTED:
+		if (bt_ag && bt_ag->disconnected) {
+			bt_ag->disconnected(ag);
+		}
+		break;
+	case BT_HFP_CONNECTING:
+		break;
+	case BT_HFP_CONFIG:
+		break;
+	case BT_HFP_CONNECTED:
+		if (bt_ag && bt_ag->connected) {
+			bt_ag->connected(ag);
+		}
+		break;
+	case BT_HFP_DISCONNECTING:
+		break;
+	default:
+		LOG_WRN("no valid (%u) state was set", state);
+		break;
+	}
+}
+
+static void bt_hfp_ag_set_call_state(struct bt_hfp_ag *ag, bt_hfp_call_state_t state)
+{
+	ag->call_state = state;
+
+	switch (state) {
+	case BT_HFP_CALL_TERMINATE:
+		atomic_clear(ag->flags);
+		k_work_cancel_delayable(&ag->deferred_work);
+		break;
+	case BT_HFP_CALL_OUTGOING:
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_OUTGOING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_INCOMING:
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_INCOMING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_ALERTING:
+		k_work_reschedule(&ag->ringing_work, K_NO_WAIT);
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_ALERTING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_ACTIVE:
+		k_work_cancel_delayable(&ag->deferred_work);
+		break;
+	case BT_HFP_CALL_HOLD:
+		break;
+	default:
+		/* Invalid state */
+		break;
+	}
+
+	if (ag->state == BT_HFP_DISCONNECTING) {
+		int err = bt_rfcomm_dlc_disconnect(&ag->rfcomm_dlc);
+
+		if (err) {
+			LOG_ERR("Fail to disconnect DLC %p", &ag->rfcomm_dlc);
+		}
+	}
+}
+
+static struct bt_ag_tx *bt_ag_tx_alloc(void)
+{
+	/* The TX context always get freed in the system workqueue,
+	 * so if we're in the same workqueue but there are no immediate
+	 * contexts available, there's no chance we'll get one by waiting.
+	 */
+	if (k_current_get() == &k_sys_work_q.thread) {
+		return k_fifo_get(&ag_tx_free, K_NO_WAIT);
+	}
+
+	if (IS_ENABLED(CONFIG_BT_HFP_AG_LOG_LEVEL_DBG)) {
+		struct bt_ag_tx *tx = k_fifo_get(&ag_tx_free, K_NO_WAIT);
+
+		if (tx) {
+			return tx;
+		}
+
+		LOG_WRN("Unable to get an immediate free bt_ag_tx");
+	}
+
+	return k_fifo_get(&ag_tx_free, K_FOREVER);
+}
+
+static void bt_ag_tx_free(struct bt_ag_tx *tx)
+{
+	LOG_DBG("Free tx buffer %p", tx);
+
+	(void)memset(tx, 0, sizeof(*tx));
+
+	k_fifo_put(&ag_tx_free, tx);
+}
+
+static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *user_data)
+{
+	struct bt_ag_tx *tx = NULL;
```

Comment:
Remove the `= NULL` - it's redundant as you're unconditionally setting the variable on the next line.

Reply:
updated